internal/transport/http2_server.go

```diff
@@ -349,10 +349,15 @@ func (t *http2Server) operateHeaders(ctx context.Context, frame *http2.MetaHeade
 		})
 		return nil
 	}
-
+	
 	if streamID%2 != 1 || streamID <= t.maxStreamID {
 		// illegal gRPC stream id.
-		return fmt.Errorf("received an illegal stream id: %v. headers frame: %+v", streamID, frame)
+		t.controlBuf.put(&goAway{
```

Comment:
This is not where I would put the goAway to the controlBuf. This would result in operateHeaders never returning an error and the err handling portion in HandleStreams never getting executed. How about we do exactly the same where t.Close(err) is called in HandleStreams and the goAway frame is created using the error returned from here?

Reply:
I am getting EOF when I read st.fr.ReadFrame(). This might how HandleStream operates on the `t.Close()`. 

When we put goAwayFrame and don't return error, the transport will close later? but if I put the goAwayFrame in the HandleStream and close the transport `t.Close()`, seems like the client is not getting GoAwayFrame, instead EOF is sent.

Looking at the operateHeader class, this is the only place where we return an error. For everything else, it was queue into `controlBuf`.  This https://github.com/grpc/grpc-go/blob/f135e982e6ca1949979c4e944942ccf0b20acc96/internal/transport/http2_server.go#L436 is also similar to what we are doing. I am in favor of going with this change. lmk

Reply:
> I am getting EOF when I read st.fr.ReadFrame() if I put GoAwayFrame in the HandleStream and then call t.Close().

Once you put the goaway Frame in HandleStreams, you should just `continue` and the next iteration of the loop would close the transport here.

 https://github.com/grpc/grpc-go/blob/f135e982e6ca1949979c4e944942ccf0b20acc96/internal/transport/http2_server.go#L639

when the next read here fails 

https://github.com/grpc/grpc-go/blob/f135e982e6ca1949979c4e944942ccf0b20acc96/internal/transport/http2_server.go#L639